API update settings, completion, chat and env files

dpsalvatierra opened the pull request:
## Describe your changes
Update Makefile for Improved

#### Added commands to set up and clean up Docker test environment.

- Modified test command to use Docker container.
- Introduced commands for Docker image deletion and test build.

#### Implement Streaming in completions.py and Other Updates:

- Added a new class and function in completions.py for streaming responses using yield and chunking
- Implemented real-time response streaming for the chat feature.
- Update .env File for Model 

#### Configuration:

- Introduced a new environment variable for model specification.
- Clarified model file placement in the project directory.
- Added notes on self-retrieving models during inference.
- Update README.md with Detailed 

#### Streaming Instructions:

- Set up variables and configuration for the OpenAI API in a local server setting.
- Detailed creating and printing responses from a ChatCompletion object with streaming.

## Checklist before requesting a review
- [x] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] I have added thorough documentation for my code.
- [ ] I have tagged PR with relevant project labels. I acknowledge that a PR without labels may be dismissed.
- [ ] If this PR addresses a bug, I have provided both a screenshot/video of the original bug and the working solution.

### Steps to Reproduce
Used completion scripts in readme and curl to verify API function

Chat Completion Streaming
```
date;curl -X 'POST'   'http://leviathan:4891/v1/chat/completions'   -H 'accept: application/json'   -H 'Content-Type: application/json'   -d @request.json
Thu Dec  7 18:25:53 EST 2023
data: {"choices": [{"delta": {"content": "\n\nAnd my question is: What are the differences between an LLC"}}]}

data: {"choices": [{"delta": {"content": " and a corporation in terms of taxation, liability, management"}}]}
```
Completion Streaming (must use openai 0.28.1)
```
 curl -X 'POST'   'http://leviathan:4891/v1/completions/'   -H 'accept: application/json'   -H 'Content-Type: application/json'   -d @completion_request.json
data: {"id": "c8fc67ac-5439-4555-bef6-309e1b3dd256", "object": "text_completion", "created": 1701991494, "model": "mistral-7b-instruct-v0.1.Q4_0", "choices": [{"text": " 5100 laptop", "index": 0, "logprobs": null, "finish_reason": ""}]}

data: {"id": "c8fc67ac-5439-4555-bef6-309e1b3dd256", "object": "text_completion", "created": 1701991494, "model": "mistral-7b-instruct-v0.1.Q4_0", "choices": [{"text": "\n\ni am looking", "index": 0, "logprobs": null, "finish_reason": ""}]}

data: {"id": "c8fc67ac-5439-4555-bef6-309e1b3dd256", "object": "text_completion", "created": 1701991494, "model": "mistral-7b-instruct-v0.1.Q4_0", "choices": [{"text": " for a new", "index": 0, "logprobs": null, "finish_reason": ""}]}

data: {"id": "c8fc67ac-5439-4555-bef6-309e1b3dd256", "object": "text_completion", "created": 1701991494, "model": "mistral-7b-instruct-v0.1.Q4_0", "choices": [{"text": " notebook that", "index": 0, "logprobs": null, "finish_reason": ""}]}

data: {"id": "c8fc67ac-5439-4555-bef6-309e1b3dd256", "object": "text_completion", "created": 1701991494, "model": "mistral-7b-instruct-v0.1.Q4_0", "choices": [{"text": " will serve", "index": 0, "logprobs": null, "finish_reason": ""}]}
```
## Notes
- Moved env file to ./gpt4all/gpt4all-api
- Removed GPU support from make for now, since it is using an external project (huggingface TGI)
- Updated Readme files, still a work in progress
- Embedding needs to be updated still

cebtenzzre commented on gpt4all-api/makefile:
```diff
@@ -1,46 +1,44 @@
-isort:
-	source $(ROOT_DIR)/venv/bin/activate; isort  --ignore-whitespace --atomic -w 120 $(APP_NAME)
\ No newline at end of file
+	sudo find . -type d -name "__pycache__" -exec rm -rf {} +
+	sudo rm -rf $(ROOT_DIR)/venv
+	sudo find $(ROOT_DIR)/$(APP_NAME) -name '*.pyc' -delete
```
There should be no need for sudo in a Makefile.

cebtenzzre commented on gpt4all-api/gpt4all_api/app/api_v1/routes/chat.py:
```diff
@@ -1,26 +1,45 @@
 import logging
 import time
-from typing import List
+from typing import List, Optional
```
It'd make more sense to follow [PEP 604](https://peps.python.org/pep-0604/) and use `Type | None` instead of `Optional[Type]`. You'll need to `from __future__ import annotations` for that.

dpsalvatierra replied:
Thanks for the heads up, I will work on this.

dpsalvatierra commented:
@cebtenzzre 
Decided to perform some improvements after your suggestions.

**Pull Request Summary:**

This pull request introduces several enhancements and updates across various files within the project. Here is a summary of the changes made:

**chat.py:**
- Added documentation.
- Utilized new annotations syntax using the `_future_` module instead of options.
- Implemented exception handling to gracefully handle errors in generating embeddings.

**completions.py:**
- Added documentation.
- Utilized new annotations syntax using the `_future_` module instead of options.
- Enhanced logic to handle different types of prompts and validate the model name.
- Implemented exception handling to gracefully handle errors in generating embeddings.
- Introduced an async generator, "event_generator," refactored from `chat.py` for consistency.

**embeddings.py:**
- Added documentation.
- Incorporated logging statements to track processing time and errors.
- Implemented exception handling to gracefully handle errors in generating embeddings.

**settings.py:**
- Temporarily commented out Huggingface GPU inference.
- Added an entry to pull the `.env` `EMBEDDING_BIN` variable.

**api.py:**
- Added an "embeddings" menu to the API Docs page.

**embeddings.py:**
- Utilized `settings.embedding` variable.

**ENV:**
- Expanded instructions and introduced the new variable "EMBEDDING_BIN."
- Added the `N_THREADS` variable to control the number of CPU threads used.

**MAKEFILE:**
- Removed the use of `sudo`.

**dockerbuild:**
- Refactored to eliminate the need for root permissions and created an "appuser" to avoid creating files as root locally.
- Addressed the issue where "__pycache__" was being created as root.

**docker-compose:**
- Added the `PATH` environment variable for the "appuser."

**README.MD**
- Updated readme file

These changes collectively improve the codebase, enhance error handling, and ensure smoother execution both locally and in containerized environments.

### Improvements:
- Better n_thread handling via env file
- Embedded is now working and added to the API Docs page
- Both Chat Completions and Completions are using the same "event_generator" for streaming
- docker container no longer runs as root, using "appuser" instead

dpsalvatierra commented:
### Refactored docker compose and dockerbuild to handle USER_ID required for fetch download of models
### Updated README file

ptrckkk commented on gpt4all-api/gpt4all_api/Dockerfile.buildkit:
```diff
@@ -1,17 +1,28 @@
-RUN --mount=type=ssh pip install -r /requirements.txt && \
-  rm -Rf /root/.cache && rm -Rf /tmp/pip-install*
+# Create a non-root user 'appuser' and switch to it
+RUN useradd -u ${USER_ID} -m appuser 
```
I tested this branch. I created a `.env` file by copying `env` with the default `USER_ID=1001`. Executing
`DOCKER_BUILDKIT=1 docker build -t gpt4all_api --progress plain -f gpt4all_api/Dockerfile.buildkit .` on ArchLinux resulted in the following error for me:

> #12 [3/6] RUN useradd -u  -m appuser
> #12 0.440 useradd: invalid user ID '-m'
> #12 ERROR: process "/bin/sh -c useradd -u ${USER_ID} -m appuser" did not complete successfully: exit code: 3

Am I doing something wrong?

dpsalvatierra replied:
Ahh I see the problem, when building with docker build, you need to pass the user 1001 variable like this
`DOCKER_BUILDKIT=1 docker build --build-arg USER_ID=1001 -t gpt4all_api --progress plain -f gpt4all_api/Dockerfile.buildkit .
`

Will update the instructions, if you perform `make testenv` or `docker-compose -f docker-compose.yaml up --build` the variable will be passed from .env

Once I update the PR, you should be able to just run `DOCKER_BUILDKIT=1 docker build $(grep 'USER_ID' .env | xargs -I {} echo --build-arg {}) -t gpt4all_api --progress plain -f gpt4all_api/Dockerfile.buildkit .`

dpsalvatierra replied:
Wait a minute, I did mention it on the new READ me :D

ptrckkk replied:
> Wait a minute, I did mention it on the new READ me

Thanks, I overlooked that. It works :+1: 

I am now wondering: With "USER_ID=1001" passed to the `docker build` command, there are two sources for the `USER_ID`: 1) the `docker build` command and 2) the `.env` file. The `USER_ID` value might not necessarily be the same.

Do both `USER_ID`s have the same meaning? Does it makes sense to apply DRY here?

dpsalvatierra replied:
Docker builds images using docker build and docker-compose. docker-compose automatically picks up environment variables, streamlining configuration management for multiple services. In contrast, docker build requires explicitly passing each environment variable as --build-arg, offering precise control over the build process but requiring more manual input.

ptrckkk commented on gpt4all-api/README.md:
````diff
@@ -1,87 +1,158 @@
 
 ```bash
-docker compose up --build gpt4all_api
+docker compose up --build -d
````
Running this command with the following `.env` file

```
MODEL_BIN="mistral-7b-openorca.Q4_0.gguf"
EMBEDDING="" # Optional, if you want to use a custom embedding, place it in the models folder and add the name here
N_THREADS=4 # 1 is default, 0 is all cores, this is an integer value not a string
USER_ID=1001 # 1001 is default, this is a string value
```

gives me the following error when POSTing chat completion requests:

> Process SpawnProcess-1:
> Traceback (most recent call last):
>   File "/usr/local/lib/python3.11/multiprocessing/process.py", line 314, in _bootstrap
>     self.run()
>   File "/usr/local/lib/python3.11/multiprocessing/process.py", line 108, in run
>     self._target(*self._args, **self._kwargs)
>   File "/usr/local/lib/python3.11/site-packages/uvicorn/_subprocess.py", line 76, in subprocess_started
>     target(sockets=sockets)
>   File "/usr/local/lib/python3.11/site-packages/uvicorn/server.py", line 59, in run
>     return asyncio.run(self.serve(sockets=sockets))
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/local/lib/python3.11/asyncio/runners.py", line 190, in run
>     return runner.run(main)
>            ^^^^^^^^^^^^^^^^
>   File "/usr/local/lib/python3.11/asyncio/runners.py", line 118, in run
>     return self._loop.run_until_complete(task)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "uvloop/loop.pyx", line 1517, in uvloop.loop.Loop.run_until_complete
>   File "/usr/local/lib/python3.11/site-packages/uvicorn/server.py", line 66, in serve
>     config.load()
>   File "/usr/local/lib/python3.11/site-packages/uvicorn/config.py", line 471, in load
>     self.loaded_app = import_from_string(self.app)
>                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/local/lib/python3.11/site-packages/uvicorn/importer.py", line 21, in import_from_string
>     module = importlib.import_module(module_str)
>              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/local/lib/python3.11/importlib/__init__.py", line 126, in import_module
>     return _bootstrap._gcd_import(name[level:], package, level)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "<frozen importlib._bootstrap>", line 1204, in _gcd_import
>   File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
>   File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
>   File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
>   File "<frozen importlib._bootstrap_external>", line 940, in exec_module
>   File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
>   File "/app/main.py", line 16, in <module>
>     from api_v1.api import router as v1_router
>   File "/app/api_v1/api.py", line 2, in <module>
>     from .routes import chat, completions, embeddings, engines, health
>   File "/app/api_v1/routes/embeddings.py", line 62, in <module>
>     embedder = Embed4All(model_path=settings.gpt4all_path)
>                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/appuser/.local/lib/python3.11/site-packages/gpt4all/gpt4all.py", line 44, in __init__
>     self.gpt4all = GPT4All(model_name or 'all-MiniLM-L6-v2-f16.gguf', n_threads=n_threads, **kwargs)
>                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/appuser/.local/lib/python3.11/site-packages/gpt4all/gpt4all.py", line 97, in __init__
>     self.config: ConfigType = self.retrieve_model(model_name, model_path=model_path, allow_download=allow_download, verbose=verbose)
>                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/appuser/.local/lib/python3.11/site-packages/gpt4all/gpt4all.py", line 187, in retrieve_model
>     config["path"] = GPT4All.download_model(model_filename, model_path, verbose=verbose, url=url)
>                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/appuser/.local/lib/python3.11/site-packages/gpt4all/gpt4all.py", line 239, in download_model
>     with open(download_path, "wb") as file, \
>          ^^^^^^^^^^^^^^^^^^^^^^^^^
> PermissionError: [Errno 13] Permission denied: '/models/all-MiniLM-L6-v2-f16.gguf'

Did I overlook some place where I should have changed "all-MiniLM-L6-v2-f16.gguf" to "mistral-7b-openorca.Q4_0.gguf"?

ptrckkk replied:
```
-> % id
uid=1000(patrick) gid=1000(patrick) groups=1000(patrick),3(sys),90(network),98(power),108(vboxusers),967(docker),991(lp),998(wheel)

-> % ls -lad gpt4all_api/models 
drwxr-xr-x 2 patrick patrick 4096  5. Jan 09:19 gpt4all_api/models

-> % ls -lad gpt4all_api/models/mistral-7b-openorca.Q4_0.gguf 
-rw-r--r-- 1 patrick patrick 4108927744  5. Jan 09:19 gpt4all_api/models/mistral-7b-openorca.Q4_0.gguf
```

dpsalvatierra replied:
Yes, I see the problem. In the instructions, I did document that you will to use the same posix ID as the host when building and running docker. The system doesn't care about the "appuser" but looks for the ID for read/write permissions on the folders which happen to be mounted from the host.

In short, use USER=1000

You could still use other users but then you need to adjust your project's posix read/write permissions on all folders to allow USER=1001 access.

ptrckkk replied:
Thank you, this solves this issue!

Using the Python code in the README file does not give me a text back, though:
```
{
  "id": "6416e937-b289-4b4c-9fc2-e89ba12c1846",
  "object": "text_completion",
  "created": 1704462349,
  "model": "mistral-7b-openorca.Q4_0",
  "choices": [
    {
      "text": "\n",
      "index": 0,
      "logprobs": -1.0,
      "finish_reason": "length"
    }
  ],
  "usage": {
    "prompt_tokens": 22,
    "completion_tokens": 1,
    "total_tokens": 23
  }
}
```

The strange thing is that '"prompt_tokens": 22` seems to reflect each char in the input prompt "Who is Michael Jordan?". I'll see if I can figure it out.

dpsalvatierra reviewed (commented):
Reviewing missing signoff for commit [0993737]
